extended-utxo-spec/extended-utxo-specification.tex

```diff
@@ -0,0 +1,1212 @@
+  those, for example share classes. \red{Eh?}
+\end{itemize}
+
+\paragraph{Performance.}
+\red{I'm just making random stuff up here.}
+The UTXO-2 and UTXO-3 models will lose some efficiency in comparison
+to the UTXO-1 model, simply because the data structures are more
+complicated.  This would even apply to transactions which only involve
+to native currency, since it would be necessary to check whether the
+\valuemap{} or \tokenmap{} contains anything that needs to be processed.  If this is
+a concern then one could implement a model with two types of
+transaction, essentially just the disjoint union of the XUTXO-1 and
+XUTXO-2 or XUTXO-3 transaction types. A simple case distinction at the
+start of a transaction could then select either a fast
+native-currency-only computation or a slower multicurrency computation.
+This would be harder to maintain though.
+
+\medskip
+A more significant cost may be that we can no longer use \verb|{-# UNPACK #-}|
+when our Value type stops being a simple combination of wrappers
+and products around primitives, but this is again an issue with any
+multi-currency proposal.
+
+  
+\subsection{Further generalisations}
+We have assumed that our basic \valuetype{} type is the type of
+natural numbers.  However, the models mostly carry over to the case
+where \valuetype{} is an arbitrary abelian group\footnotemark.  We can
+even replace \valuetype{} with a commutative monoid as long as we
+relinquish the ability to destroy quantities of currencies (since we
+can no longer subtract in general). Whether these generalisations are
+useful in practice remains to be seen.
+
+\footnotetext{If we allow a non-abelian group then we may be in
+  trouble.  If we forge $a$ units of currency, then forge $b$ units,
+  then destroy $a$ units again, we're left with $aba^{-1}$ (or
+  $a+b-a$) units, which won't be equal to $b$ in general: if your
+  currency consists of elements of the free group on $\{a,b\}$ then
+  things could get pretty interesting.  Similar things may happen if
+  we need to split up an output to get some change. In the monoid
+  case, we might get unexpected cancellation as well: $a+b=0$, but
+  $a,b \ne 0$: consider $\mathbb{Z}_{24}$ or something. I suppose
+  doing things like this would let us write weird papers to impress
+  the cryptocurrency community. It could still be useful for something
+  though...}
+
+
+
+\section{What else do we need to worry about?}
+\begin{itemize}
+
+\item It might be worth adding an appendix that contains the entire
+  description of XUTXO-3 plus the relevant types and so on in one place.
+  
+\item Should we mention the problem of \textbf{validating the next data script}?
+%% https://github.com/input-output-hk/plutus/issues/426
+
+\item Sometimes I'm saying ``datascript'' and at other times ``data script''.
+  Maybe we should try to be consistent about this.
+  
+\item Do we need to worry about who signs the various scripts, or is that
+  something the wallet deals with?
```

Comment:
I think we should leave out the signatures if possible (unless they are really necessary for the examples, which I'm not 100% sure about yet)

Reply:
The signature stuff will have to be specified somewhere, and I wasn't sure where. We should check with Manuel (and maybe I did, but I don't recall if I got an answer).

Reply:
It should go in the real ledger spec. The point of this spec is EUTxO, we should avoid talking too much about stuff that's not directly relevant to that.